src/test_s2c_srv.c

```diff
@@ -256,89 +259,90 @@ int test_s2c(int ctlsockfd, tcp_stat_agent* agent, TestOptions* testOptions,
       // Proceed.
-      // Note the new socket fd - xmitfd - used in the throughput test
-ximfd: xmitsfd = accept(testOptions->s2csockfd,
-                        (struct sockaddr *) &cli_addr, &clilen);
-       if (xmitsfd > 0) {
-         log_println(6, "accept() for %d completed",
-                     testOptions->child0);
-         procstatusenum = PROCESS_STARTED;
-         proctypeenum = CONNECT_TYPE;
-         protolog_procstatus(testOptions->child0, testids, proctypeenum,
-                             procstatusenum, xmitsfd);
-         if (testOptions->connection_flags & WEBSOCKET_SUPPORT) {
-	   // To preserve user privacy, make sure that the HTTP header
-	   // processing is done prior to the start of packet capture, as many
-	   // browsers have headers that uniquely identitfy a single user.
-           if (initialize_websocket_connection(xmitsfd, 0, "s2c") != 0) {
-             xmitsfd = 0;
-           } 
-         } 
-         break;
-       }
-       // socket interrupted, wait some more
-       if ((xmitsfd == -1) && (errno == EINTR)) {
-         log_println(
-             6,
-             "Child %d interrupted while waiting for accept() to complete",
-             testOptions->child0);
-         goto ximfd;
-       }
-       log_println(
-           6,
-           "-------     S2C connection setup for %d returned because (%d)",
-           testOptions->child0, errno);
-       if (xmitsfd < 0)  // other socket errors, quit
-         return -errno;
-       if (++j == 4) {  // retry exceeded, quit
-         log_println(
-             6,
-             "s2c child %d, unable to open connection, return from test",
-             testOptions->child0);
-         return -102;
-       }
+      // Note the new connection - s2c_conn - used in the throughput test
+    ximfd:
+      s2c_conn.socket = accept(testOptions->s2csockfd,
+                               (struct sockaddr *)&cli_addr, &clilen);
+      if (s2c_conn.socket > 0) {
+        log_println(6, "accept() for %d completed", testOptions->child0);
+        procstatusenum = PROCESS_STARTED;
+        proctypeenum = CONNECT_TYPE;
+        protolog_procstatus(testOptions->child0, testids, proctypeenum,
+                            procstatusenum, s2c_conn.socket);
+        if (testOptions->connection_flags & TLS_SUPPORT) {
+          s2c_conn.ssl = SSL_new(ctx);
+          if (s2c_conn.ssl == NULL) return ENOMEM;
+          if (SSL_set_fd(s2c_conn.ssl, s2c_conn.socket) == 0) return EIO;
+          if (SSL_accept(s2c_conn.ssl) != 1) return EIO;
+        }
+        if (testOptions->connection_flags & WEBSOCKET_SUPPORT) {
+          // To preserve user privacy, make sure that the HTTP header
+          // processing is done prior to the start of packet capture, as many
+          // browsers have headers that uniquely identitfy a single user.
+          if (initialize_websocket_connection(&s2c_conn, 0, "s2c") != 0) {
+            s2c_conn.socket = 0;
+          } 
```

Comment:
Doesn't this leak `s2c_conn.ssl `? And same question about whether we should be closing the socket too.